jax/interpreters/pxla.py

```diff
@@ -331,32 +384,42 @@ class ShardedDeviceArray(xla.DeviceArray):
   behavior of an ndarray so that it can be treated by user code as an ndarray;
   that is, it is only an optimization to reduce transfers.
 
-  The number of device buffers underlying a ShardedDeviceArray instance is equal
-  to the number of replicas of the computation that produced it. Each buffer
-  represents a shard of the original array, meaning a slice along its leading
-  axis. These component buffers reside on distinct devices, but need not
-  represent distinct logical shards. The correspondence can be computed with
-  the assign_shards_to_replicas function.
+  Attributes:
+    aval: A ShapedArray indicating the shape and dtype of this array.
+    sharding_spec: describes how this array is sharded across `device_buffers`.
+    device_buffers: the buffers containing the data for this array. Each buffer
+      is the same shape and on a different device. Buffers are in row-major
+      order, with replication treated as an extra innermost dimension. For
+      example, if the unsharded data is [1, 2], the replicated data should be
+      laid out here as [1, 1, 2, 2].
+    indices: the result of spec_to_indices(sharding_spec). Can optionally be
+      precomputed for efficiency. A list the same length as `device_buffers`
+      (equal to `prod(sharding_spec.shards_per_axis)`). Each index indicates what
+      portion of the full array is stored in the corresponding device buffer,
+      i.e. `array[indices[i]] == device_buffers[i].to_py()`.
   """
-  __slots__ = ["device_buffers", "axis_size"]
-  _collect = staticmethod(onp.stack)
-
-  def __init__(self, aval, device_buffers):
+  __slots__ = ["device_buffers", "sharding_spec", "indices"]
+
+  # TODO(skye): expose PyLocalBuffers in xla_client
+  def __init__(self,
+               aval: ShapedArray,
+               sharding_spec: ShardingSpec,
+               device_buffers: List[xb.xla_client._xla.PyLocalBuffer],
+               indices: List[Index] = None):
+    # TODO(skye): assert invariants. Keep performance in mind though.
```

Comment:
We can put assertions under some global testing flags, like core.check_jaxpr

Reply:
Agreed, I'm gonna leave it as a TODO for now though. All the unit tests pass so it can't be too off right? :)